File: apex-brain/tools/orchestrator.py
#!/usr/bin/env python3
"""Workflow orchestrator. Reads workflow markdown files, executes steps via Claude API."""

import argparse
import json
import os
import re
import subprocess
import sys
import urllib.request
import urllib.error
from datetime import datetime
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent

# Auto-load .env file if present
_env_file = REPO_ROOT / ".env"
if _env_file.exists():
    for line in _env_file.read_text().splitlines():
        line = line.strip()
        if line and not line.startswith("#") and "=" in line:
            key, _, value = line.partition("=")
            if value and not os.environ.get(key.strip()):
                os.environ[key.strip()] = value.strip()

WORKFLOWS_DIR = REPO_ROOT / "workflows"
OUTPUTS_DIR = REPO_ROOT / "outputs"
ENTITIES_DIR = REPO_ROOT / "entities"
CONTEXT_DIR = REPO_ROOT / "context"
SKILLS_DIR = REPO_ROOT / "skills"
RUN_LOG = REPO_ROOT / "decisions" / "run_log.jsonl"

CLAUDE_API_URL = "https://api.anthropic.com/v1/messages"
CLAUDE_MODEL = "claude-sonnet-4-20250514"
MAX_TOKENS = 4096

# Tools that can be dispatched from workflow steps
TOOL_MAP = {
    "post_to_linkedin": "tools/post_to_linkedin.py",
    "post_to_twitter": "tools/post_to_twitter.py",
    "fetch_analytics": "tools/fetch_analytics.py",
    "scrape_trending": "tools/scrape_trending.py",
    "send_email": "tools/send_email.py",
    "deploy_to_vercel": "tools/deploy_to_vercel.py",
    "browser_agent": "tools/browser_agent.py",
}


def load_file(path: Path) -> str:
    """Read a file, return empty string if missing."""
    try:
        return path.read_text()
    except FileNotFoundError:
        return ""


def build_entity_context(entity: str) -> str:
    """Load config.md, brand.md, goals.md for an entity."""
    entity_dir = ENTITIES_DIR / entity
    parts = []
    for fname in ["config.md", "brand.md", "goals.md"]:
        content = load_file(entity_dir / fname)
        if content:
            parts.append(content)
    return "\n\n---\n\n".join(parts)


def build_skill_context(workflow_text: str) -> str:
    """Extract skill names from workflow and load their SKILL.md files."""
    skills_used = re.findall(r"[-*]\s*(\S+)", re.search(r"## Skills Used\n(.*?)(?:\n##|\Z)", workflow_text, re.DOTALL).group(1)) if "## Skills Used" in workflow_text else []
    parts = []
    for skill in skills_used:
        skill_file = SKILLS_DIR / skill / "SKILL.md"
        content = load_file(skill_file)
        if content:
            parts.append(f"### Skill: {skill}\n{content}")
    return "\n\n".join(parts)


def call_claude(system_prompt: str, user_prompt: str, api_key: str) -> dict:
    """Call the Anthropic Messages API."""
    payload = {
        "model": CLAUDE_MODEL,
        "max_tokens": MAX_TOKENS,
        "system": system_prompt,
        "messages": [{"role": "user", "content": user_prompt}],
    }
    body = json.dumps(payload).encode()
    req = urllib.request.Request(CLAUDE_API_URL, data=body, headers={
        "x-api-key": api_key,
        "anthropic-version": "2023-06-01",
        "content-type": "application/json",
    })
    try:
        with urllib.request.urlopen(req, timeout=120) as resp:
            data = json.loads(resp.read())
        text = data.get("content", [{}])[0].get("text", "")
        return {"status": "ok", "text": text, "usage": data.get("usage", {})}
    except urllib.error.HTTPError as e:
        return {"status": "error", "code": e.code, "message": e.read().decode()}
    except Exception as e:
        return {"status": "error", "message": str(e)}


def run_tool(tool_name: str, args: list, dry_run: bool = False) -> dict:
    """Execute a Python tool script and capture JSON output."""
    script = REPO_ROOT / TOOL_MAP.get(tool_name, "")
    if not script.exists():
        return {"status": "error", "message": f"Tool not found: {tool_name}"}

    cmd = [sys.executable, str(script)] + args
    if dry_run:
        cmd.append("--dry-run")

    try:
        result = subprocess.run(cmd, capture_output=True, text=True, timeout=60)
        try:
            return json.loads(result.stdout)
        except json.JSONDecodeError:
            return {"status": "ok", "raw_output": result.stdout, "stderr": result.stderr}
    except subprocess.TimeoutExpired:
        return {"status": "error", "message": f"Tool {tool_name} timed out"}


def parse_workflow(path: Path) -> dict:
    """Extract metadata from a workflow markdown file."""
    text = path.read_text()
    wf = {"path": str(path), "entity": path.parent.name, "name": path.stem, "steps": [], "schedule": None, "raw": text}

    sched = re.search(r"schedule:\s*(.+)", text, re.IGNORECASE)
    if sched:
        wf["schedule"] = sched.group(1).strip()

    for match in re.finditer(r"(?:^##\s*Step\s*\d+[:\s]*(.+)|^\d+\.\s*\*\*(.+?)\*\*)", text, re.MULTILINE):
        step_name = (match.group(1) or match.group(2)).strip()
        wf["steps"].append(step_name)

    return wf


def discover_workflows(schedule_filter: str = None) -> list:
    """Find all workflow files, optionally filtered by schedule keyword."""
    workflows = []
    if not WORKFLOWS_DIR.exists():
        return workflows
    for md in WORKFLOWS_DIR.rglob("*.md"):
        wf = parse_workflow(md)
        if schedule_filter and wf["schedule"]:
            if schedule_filter.lower() not in wf["schedule"].lower():
                continue
        elif schedule_filter and not wf["schedule"]:
            continue
        workflows.append(wf)
    return workflows


def execute_workflow(wf: dict, api_key: str, dry_run: bool = False) -> dict:
    """Execute a workflow by sending each step to Claude with full context."""
    result = {
        "workflow": wf["name"],
        "entity": wf["entity"],
        "steps": [],
        "timestamp": datetime.utcnow().isoformat(),
    }

    # Build context
    entity_context = build_entity_context(wf["entity"])
    skill_context = build_skill_context(wf["raw"])
    global_context = load_file(CONTEXT_DIR / "priorities.md")

    system_prompt = f"""You are the Apex Brain orchestrator for the entity "{wf['entity']}".
Your job is to execute one step of a workflow and produce concrete output.

## Entity Context
{entity_context}

## Current Priorities
{global_context}

## Available Skills
{skill_context}

## Rules
- Produce the actual deliverable for each step (draft text, research findings, etc.)
- Return valid JSON with keys: "output" (the content), "tool_calls" (list of tools to invoke, or empty), "needs_approval" (bool), "confidence" (0-100)
- For tool_calls, use format: {{"tool": "tool_name", "args": ["--flag", "value"]}}
- If a step produces content for human review, set needs_approval: true
- Be direct, no filler. Match the entity's brand voice."""

    # Accumulate step outputs so later steps have context from earlier ones
    conversation_so_far = []

    for step in wf["steps"]:
        if dry_run:
            result["steps"].append({"step": step, "status": "dry_run"})
            continue

        if not api_key:
            result["steps"].append({"step": step, "status": "skipped", "reason": "no API key"})
            continue

        # Build user prompt with conversation history
        history = ""
        if conversation_so_far:
            history = "## Previous steps completed:\n" + "\n".join(
                f"- **{s['step']}**: {s.get('output_preview', 'done')}" for s in conversation_so_far
            ) + "\n\n"

        user_prompt = f"""{history}## Current Step: {step}

Full workflow for reference:
{wf['raw']}

Execute this step now. Return JSON only."""

        claude_result = call_claude(system_prompt, user_prompt, api_key)

        step_result = {"step": step, "status": "completed"}

        if claude_result["status"] == "ok":
            # Try to parse Claude's JSON response
            raw_text = claude_result["text"]
            try:
                # Extract JSON from possible markdown code fences
                json_match = re.search(r"```(?:json)?\s*(\{.*?\})\s*```", raw_text, re.DOTALL)
                parsed = json.loads(json_match.group(1) if json_match else raw_text)
                step_result["output"] = parsed.get("output", raw_text)
                step_result["needs_approval"] = parsed.get("needs_approval", False)
                step_result["confidence"] = parsed.get("confidence", 50)

                # Dispatch any tool calls Claude requested
                for tc in parsed.get("tool_calls", []):
                    tool_name = tc.get("tool", "")
                    tool_args = tc.get("args", [])
                    if tool_name in TOOL_MAP:
                        tool_result = run_tool(tool_name, tool_args, dry_run)
                        step_result.setdefault("tool_results", []).append({
                            "tool": tool_name,
                            "result": tool_result,
                        })
            except (json.JSONDecodeError, AttributeError):
                step_result["output"] = raw_text
                step_result["needs_approval"] = True
                step_result["confidence"] = 50

            step_result["usage"] = claude_result.get("usage", {})
        else:
            step_result["status"] = "error"
            step_result["error"] = claude_result

        result["steps"].append(step_result)
        conversation_so_far.append({
            "step": step,
            "output_preview": str(step_result.get("output", ""))[:200],
        })

    # Save outputs
    entity_out = OUTPUTS_DIR / wf["entity"]
    entity_out.mkdir(parents=True, exist_ok=True)
    output_file = entity_out / f"{wf['name']}_{datetime.utcnow().strftime('%Y%m%d_%H%M%S')}.json"
    output_file.write_text(json.dumps(result, indent=2))

    # Log the run
    RUN_LOG.parent.mkdir(parents=True, exist_ok=True)
    with open(RUN_LOG, "a") as f:
        f.write(json.dumps({
            "workflow": wf["name"],
            "entity": wf["entity"],
            "timestamp": result["timestamp"],
            "steps_completed": sum(1 for s in result["steps"] if s["status"] == "completed"),
            "steps_total": len(result["steps"]),
            "needs_approval": any(s.get("needs_approval") for s in result["steps"]),
        }) + "\n")

    return result


def main():
    parser = argparse.ArgumentParser(description="Apex Brain workflow orchestrator")
    parser.add_argument("--schedule", help="Filter by schedule (daily, weekly, etc.)")
    parser.add_argument("--entity", help="Run only workflows for this entity")
    parser.add_argument("--workflow", help="Run a specific workflow by name")
    parser.add_argument("--list", action="store_true", help="List workflows without running")
    parser.add_argument("--dry-run", action="store_true")
    args = parser.parse_args()

    api_key = os.environ.get("ANTHROPIC_API_KEY")

    workflows = discover_workflows(args.schedule)
    if args.entity:
        workflows = [w for w in workflows if w["entity"] == args.entity]
    if args.workflow:
        workflows = [w for w in workflows if w["name"] == args.workflow]

    if args.list:
        # Strip raw content for listing
        for w in workflows:
            w.pop("raw", None)
        print(json.dumps(workflows, indent=2))
        return

    if not workflows:
        print(json.dumps({"status": "no_workflows", "filter": {"schedule": args.schedule, "entity": args.entity}}))
        return

    if not api_key and not args.dry_run:
        print(json.dumps({"status": "error", "message": "ANTHROPIC_API_KEY not set. Use --dry-run to test without it."}))
        sys.exit(1)

    results = []
    for wf in workflows:
        result = execute_workflow(wf, api_key, args.dry_run)
        results.append(result)

    print(json.dumps(results, indent=2))


if __name__ == "__main__":
    main()

File: automations/src/trigger/brain/orchestrator.ts
import { task } from "@trigger.dev/sdk";
import Anthropic from "@anthropic-ai/sdk";
import * as fs from "fs";
import * as path from "path";
import { BRAIN_ROOT } from "../outputs/config.js";

const CLAUDE_MODEL = "claude-sonnet-4-20250514";
const MAX_TOKENS = 4096;

const WORKFLOWS_DIR = path.join(BRAIN_ROOT, "workflows");
const OUTPUTS_DIR = path.join(BRAIN_ROOT, "outputs");
const ENTITIES_DIR = path.join(BRAIN_ROOT, "entities");
const CONTEXT_DIR = path.join(BRAIN_ROOT, "context");
const SKILLS_DIR = path.join(BRAIN_ROOT, "skills");
const RUN_LOG = path.join(BRAIN_ROOT, "decisions", "run_log.jsonl");

// --- Helpers ---

function loadFile(filePath: string): string {
  try {
    return fs.readFileSync(filePath, "utf-8");
  } catch {
    return "";
  }
}

function buildEntityContext(entity: string): string {
  const entityDir = path.join(ENTITIES_DIR, entity);
  const parts: string[] = [];
  for (const fname of ["config.md", "brand.md", "goals.md"]) {
    const content = loadFile(path.join(entityDir, fname));
    if (content) parts.push(content);
  }
  return parts.join("\n\n---\n\n");
}

function buildSkillContext(workflowText: string): string {
  const skillsMatch = workflowText.match(
    /## Skills Used\n([\s\S]*?)(?:\n##|$)/
  );
  if (!skillsMatch) return "";

  const skillNames = [...skillsMatch[1].matchAll(/[-*]\s*(\S+)/g)].map(
    (m) => m[1]
  );
  const parts: string[] = [];
  for (const skill of skillNames) {
    const content = loadFile(path.join(SKILLS_DIR, skill, "SKILL.md"));
    if (content) parts.push(`### Skill: ${skill}\n${content}`);
  }
  return parts.join("\n\n");
}

interface ParsedWorkflow {
  path: string;
  entity: string;
  name: string;
  steps: string[];
  schedule: string | null;
  raw: string;
}

function parseWorkflow(filePath: string): ParsedWorkflow {
  const text = fs.readFileSync(filePath, "utf-8");
  const wf: ParsedWorkflow = {
    path: filePath,
    entity: path.basename(path.dirname(filePath)),
    name: path.basename(filePath, ".md"),
    steps: [],
    schedule: null,
    raw: text,
  };

  const schedMatch = text.match(/schedule:\s*(.+)/i);
  if (schedMatch) wf.schedule = schedMatch[1].trim();

  const stepPattern =
    /(?:^##\s*Step\s*\d+[:\s]*(.+)|^\d+\.\s*\*\*(.+?)\*\*)/gm;
  let match: RegExpExecArray | null;
  while ((match = stepPattern.exec(text)) !== null) {
    wf.steps.push((match[1] || match[2]).trim());
  }

  return wf;
}

function discoverWorkflows(scheduleFilter?: string): ParsedWorkflow[] {
  if (!fs.existsSync(WORKFLOWS_DIR)) return [];

  const workflows: ParsedWorkflow[] = [];

  function walk(dir: string) {
    for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        walk(full);
      } else if (entry.name.endsWith(".md")) {
        const wf = parseWorkflow(full);
        if (scheduleFilter) {
          if (
            wf.schedule &&
            wf.schedule.toLowerCase().includes(scheduleFilter.toLowerCase())
          ) {
            workflows.push(wf);
          }
        } else {
          workflows.push(wf);
        }
      }
    }
  }

  walk(WORKFLOWS_DIR);
  return workflows;
}

interface StepResult {
  step: string;
  status: string;
  output?: any;
  needs_approval?: boolean;
  confidence?: number;
  usage?: any;
  error?: any;
}

interface WorkflowResult {
  workflow: string;
  entity: string;
  steps: StepResult[];
  timestamp: string;
}

async function executeWorkflow(
  wf: ParsedWorkflow,
  claude: Anthropic
): Promise<WorkflowResult> {
  const result: WorkflowResult = {
    workflow: wf.name,
    entity: wf.entity,
    steps: [],
    timestamp: new Date().toISOString(),
  };

  const entityContext = buildEntityContext(wf.entity);
  const skillContext = buildSkillContext(wf.raw);
  const globalContext = loadFile(path.join(CONTEXT_DIR, "priorities.md"));

  const systemPrompt = `You are the Apex Brain orchestrator for the entity "${wf.entity}".
Your job is to execute one step of a workflow and produce concrete output.

## Entity Context
${entityContext}

## Current Priorities
${globalContext}

## Available Skills
${skillContext}

## Rules
- Produce the actual deliverable for each step (draft text, research findings, etc.)
- Return valid JSON with keys: "output" (the content), "needs_approval" (bool), "confidence" (0-100)
- If a step produces content for human review, set needs_approval: true
- Be direct, no filler. Match the entity's brand voice.`;

  const conversationSoFar: { step: string; output_preview: string }[] = [];

  for (const step of wf.steps) {
    let history = "";
    if (conversationSoFar.length > 0) {
      history =
        "## Previous steps completed:\n" +
        conversationSoFar
          .map((s) => `- **${s.step}**: ${s.output_preview}`)
          .join("\n") +
        "\n\n";
    }

    const userPrompt = `${history}## Current Step: ${step}

Full workflow for reference:
${wf.raw}

Execute this step now. Return JSON only.`;

    try {
      const response = await claude.messages.create({
        model: CLAUDE_MODEL,
        max_tokens: MAX_TOKENS,
        system: systemPrompt,
        messages: [{ role: "user", content: userPrompt }],
      });

      const rawText =
        response.content[0].type === "text" ? response.content[0].text : "";

      const stepResult: StepResult = { step, status: "completed" };

      try {
        const jsonMatch = rawText.match(
          /```(?:json)?\s*(\{[\s\S]*?\})\s*```/
        );
        const parsed = JSON.parse(jsonMatch ? jsonMatch[1] : rawText);
        stepResult.output = parsed.output || rawText;
        stepResult.needs_approval = parsed.needs_approval ?? false;
        stepResult.confidence = parsed.confidence ?? 50;
      } catch {
        stepResult.output = rawText;
        stepResult.needs_approval = true;
        stepResult.confidence = 50;
      }

      stepResult.usage = response.usage;
      result.steps.push(stepResult);
      conversationSoFar.push({
        step,
        output_preview: String(stepResult.output).slice(0, 200),
      });
    } catch (err: any) {
      result.steps.push({
        step,
        status: "error",
        error: err.message || String(err),
      });
    }
  }

  // Save outputs
  const entityOut = path.join(OUTPUTS_DIR, wf.entity);
  fs.mkdirSync(entityOut, { recursive: true });
  const ts = new Date().toISOString().replace(/[:.]/g, "-");
  const outputFile = path.join(entityOut, `${wf.name}_${ts}.json`);
  fs.writeFileSync(outputFile, JSON.stringify(result, null, 2));

  // Log the run
  fs.mkdirSync(path.dirname(RUN_LOG), { recursive: true });
  fs.appendFileSync(
    RUN_LOG,
    JSON.stringify({
      workflow: wf.name,
      entity: wf.entity,
      timestamp: result.timestamp,
      steps_completed: result.steps.filter((s) => s.status === "completed")
        .length,
      steps_total: result.steps.length,
      needs_approval: result.steps.some((s) => s.needs_approval),
    }) + "\n"
  );

  return result;
}

// --- Main task ---

export const runOrchestrator = task({
  id: "run-orchestrator",
  maxDuration: 300,
  run: async (payload: {
    schedule?: string;
    entity?: string;
    workflow?: string;
  }) => {
    const claude = new Anthropic({ apiKey: process.env.ANTHROPIC_API_KEY });

    let workflows = discoverWorkflows(payload.schedule);

    if (payload.entity) {
      workflows = workflows.filter((w) => w.entity === payload.entity);
    }
    if (payload.workflow) {
      workflows = workflows.filter((w) => w.name === payload.workflow);
    }

    if (workflows.length === 0) {
      console.log("No workflows matched filters", payload);
      return { status: "no_workflows", filters: payload };
    }

    console.log(
      `Running ${workflows.length} workflow(s): ${workflows.map((w) => `${w.entity}/${w.name}`).join(", ")}`
    );

    const results: WorkflowResult[] = [];
    for (const wf of workflows) {
      const result = await executeWorkflow(wf, claude);
      results.push(result);
    }

    return results;
  },
});
